SignalServiceKit/src/Messages/OWSMessageReceiver.m

```diff
@@ -222,43 +274,95 @@ - (void)enqueueEnvelopeForProcessing:(OWSSignalServiceProtosEnvelope *)envelope
+            // We don't want to process incoming messages until database
+            // view registration is complete.
+            return;
+        }
+
+        if (self.isDrainingQueue) {
+            return;
+        }
+        self.isDrainingQueue = YES;
+
+        [self drainQueueWorkStep];
+    });
 }
 
 - (void)drainQueueWorkStep
 {
     AssertIsOnMainThread();
 
-    OWSMessageProcessingJob *_Nullable job = [self.finder nextJob];
-    if (job == nil) {
+    NSArray<OWSMessageProcessingJob *> *jobs = [self.finder nextJobsForBatchSize:kIncomingMessageBatchSize];
+    OWSAssert(jobs);
+    if (jobs.count < 1) {
         self.isDrainingQueue = NO;
-        DDLogVerbose(@"%@ Queue is drained", self.tag);
+        DDLogVerbose(@"%@ Queue is drained.", self.tag);
         return;
     }
 
-    [self processJob:job
-          completion:^{
-              DDLogVerbose(@"%@ completed job. %lu jobs left.",
-                  self.tag,
-                  (unsigned long)[OWSMessageProcessingJob numberOfKeysInCollection]);
-              [self drainQueueWorkStep];
-          }];
+    [self processJobs:jobs
+           completion:^{
+               dispatch_async(dispatch_get_main_queue(), ^{
+                   for (OWSMessageProcessingJob *job in jobs) {
+                       [self.finder removeJobWithId:job.uniqueId];
+                   }
+                   DDLogVerbose(@"%@ completed %zd jobs. %zd jobs left.",
+                       self.tag,
+                       jobs.count,
+                       [OWSMessageProcessingJob numberOfKeysInCollection]);
+                   [self drainQueueWorkStep];
+               });
+           }];
 }
 
-- (void)processJob:(OWSMessageProcessingJob *)job completion:(void (^)())completion
+- (void)processJobs:(NSArray<OWSMessageProcessingJob *> *)jobs completion:(void (^)())completion
 {
-    [self.messagesManager processEnvelope:job.envelopeProto
-                               completion:^{
-                                   [self.finder removeJobWithId:job.uniqueId];
-                                   completion();
-                               }];
+    [self processJobs:jobs
+         unprocessedJobs:[jobs mutableCopy]
+        plaintextDataMap:[NSMutableDictionary new]
+              completion:completion];
+}
+
+- (void)processJobs:(NSArray<OWSMessageProcessingJob *> *)jobs
+     unprocessedJobs:(NSMutableArray<OWSMessageProcessingJob *> *)unprocessedJobs
+    plaintextDataMap:(NSMutableDictionary<NSString *, NSData *> *)plaintextDataMap
+          completion:(void (^)())completion
+{
+    OWSAssert(jobs.count > 0);
+    OWSAssert(unprocessedJobs.count <= jobs.count);
+
+    if (unprocessedJobs.count < 1) {
+        for (OWSMessageProcessingJob *job in jobs) {
+            NSData *_Nullable plaintextData = plaintextDataMap[job.uniqueId];
+            [self.batchMessageProcessor enqueueEnvelopeData:job.envelopeData plaintextData:plaintextData];
```

Comment:
Current batch size is 10, current max envelope size is 250k, decrypted size should be smaller than encrypted size, so max batch size in memory < 2.5 MB.  I thought that was reasonable?  There are some perf benefits to this arrangement.

Reply:
Note that most batches will be just 1-2 messages and batches should be short-lived in memory.  If we're doing large batches its because "we're slammed" for one reason or another and perf will matter.

Reply:
My concern actually wasn't perf - just about code clarify. From a strictly logical standpoint it seems unnecessary. Instead of sticking them in a data structure and enqueuing them later, why don't we just immediately enqueue them?

Is it related to this comment?:

+// We need to use a consistent batch size throughout
 +// the incoming message pipeline (i.e. in the
 +// "decrypt" and "process" steps), or the pipeline
 +// doesn't flow smoothly.

e.g. I can imagine if we had 10 jobs, and handed them off individually as we finished them, the downstream work queue would start with a batch size of 1, and then continue with a batch of 9, which may be be less efficient than waiting a little bit to do a single batch of 10, as you are here.

...but that is just my speculation and correlation across the whole PR - the motivation is not obvious from reading this code locally.